src/extensions/default/PhpTooling/main.js

```diff
@@ -0,0 +1,192 @@
+        DefaultEventHandlers = brackets.getModule("languageTools/DefaultEventHandlers"),
+        PreferencesManager  = brackets.getModule("preferences/PreferencesManager"),
+        Strings             = brackets.getModule("strings"),
+        Dialogs             = brackets.getModule("widgets/Dialogs"),
+        DefaultDialogs      = brackets.getModule("widgets/DefaultDialogs"),
+        Commands               = brackets.getModule("command/Commands"),
+        CommandManager         = brackets.getModule("command/CommandManager"),
+        StringUtils             = brackets.getModule("utils/StringUtils");
+
+    var clientFilePath = ExtensionUtils.getModulePath(module, "client.js"),
+        clientName = "PhpClient",
+        _client = null,
+        phpConfig = {},
+        DEBUG_OPEN_PREFERENCES_IN_SPLIT_VIEW  = "debug.openPrefsInSplitView";
+
+    PreferencesManager.definePreference("php", "object", {
+        enablePhpTooling: true,
+        showDiagnosisOnType: false,
+        executablePath: "php",
+        validate_executablePath: "php",
+        memoryLimit: "default"
+    }, {
+        description: Strings.DESCRIPTION_PHP_TOOLING_CONFIGURATION
+    });
+
+    PreferencesManager.on("change", "php", function () {
+        var newPhpConfig = PreferencesManager.get("php");
+
+        if((newPhpConfig["executablePath"] !== phpConfig["executablePath"])
+            || (newPhpConfig["enablePhpTooling"] !== phpConfig["enablePhpTooling"]))  {
+            phpConfig = newPhpConfig;
+            startPhpServer();
+            return;
+        }
+        phpConfig = newPhpConfig;
+    });
+
+    var handleProjectOpen = function (event, directory) {
+        _client.stop()
+            .done(function (){
+                setTimeout(function () {
+                    _client.start({
+                        rootPath: directory.fullPath
+                    });
+                }, 1500);
+            });
+    };
+
+    function registerToolingProviders(client) {
+        var languageClient = client || _client;
+        var chProvider = new DefaultProviders.CodeHintsProvider(languageClient),
+            phProvider = new DefaultProviders.ParameterHintsProvider(languageClient),
+            jdProvider = new DefaultProviders.JumpToDefProvider(languageClient),
+            lProvider = new DefaultProviders.LintingProvider(languageClient);
+
+        CodeHintManager.registerHintProvider(chProvider, ["php"], 0);
+        ParameterHintManager.registerHintProvider(phProvider, ["php"], 0);
+        JumpToDefManager.registerJumpToDefProvider(jdProvider, ["php"], 0);
+        CodeInspection.register(["php"], {
+            name: "Diagnostics",
+            scanFile: lProvider.getInspectionResults.bind(lProvider)
+        });
+
+        languageClient.addOnDiagnostics(lProvider.setInspectionResults.bind(lProvider));
+    }
+
+    function startPhpServer() {
+        if(_client && phpConfig["enablePhpTooling"]) {
```

Comment:
Two suggestions:

- Decouple the validation logic from this function. You can do it on App Ready and then prompt the UI on opening a PHP file the first time. Save that to state.

- I know it seems like a nice optimization to start the server on first open of PHP file, but it doesn't seem like valid enforcement if the extension itself enforces it. Since this is a reference implementation, we'd want to provide the developer with a way where they can defer the activation of the server if they want from the brackets core itself. This is a nice optimization which other LSP extension can leverage as well.

Reply:
I think we should always validate Server config before starting. 
otherwise if server start fails for any reason (e.g user set wrong executable Path), we wont be able to notify user about failure.

Reply:
It's up to you. I find the current way a bit clumsy.